internal/pkg/service/buffer/migration/migration.go

```diff
@@ -0,0 +1,219 @@
+package migration
+
+import (
+	"context"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/c2h5oh/datasize"
+	"github.com/umisama/go-regexpcache"
+	etcd "go.etcd.io/etcd/client/v3"
+	"go.etcd.io/etcd/client/v3/concurrency"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/log"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/buffer/worker/distribution"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/iterator"
+	"github.com/keboola/keboola-as-code/internal/pkg/service/common/etcdop/op"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils/errors"
+)
+
+const (
+	migrationLockKey = etcdop.Key("runtime/migration/20230418/lock")
+	migrationDoneKey = etcdop.Key("runtime/migration/20230418/done")
+	maxTxnKeys       = 40
+	maxTxnSize       = 800 * datasize.KB
+)
+
+func APINode(logger log.Logger, client *etcd.Client) error {
+	if err := waitForMigration(logger, client); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func WorkerNode(logger log.Logger, client *etcd.Client, dist *distribution.Node) error {
+	sess, err := concurrency.NewSession(client)
+	if err != nil {
+		return err
+	}
+
+	// Wait for all new nodes
+	logger.Info("waiting 3 minutes")
+	<-time.After(3 * time.Minute)
+
+	// Only one worker should do migration
+	if !dist.MustCheckIsOwner("migration") {
+		return waitForMigration(logger, client)
+	}
+	ok, err := migrationLockKey.PutIfNotExists("locked", etcd.WithLease(sess.Lease())).Do(context.Background(), client)
+	if err != nil {
+		return err
+	}
+	if !ok {
+		logger.Warn("migration is already running")
+		return waitForMigration(logger, client)
+	}
+
+	return doMigration(logger, client, sess)
+}
+
+func doMigration(logger log.Logger, client *etcd.Client, sess *concurrency.Session) error {
+	logger = logger.AddPrefix("[migration]")
```

Comment:
na jednom worker pobezi migracia, aby to bolo rychlejsie tak spajame viacero klucov do jednej transakcie